R/tmb-methods.R

```diff
@@ -427,3 +427,55 @@ h_residuals_response <- function(object) {
   assert_class(object, "mmrm_tmb")
   component(object, "y_vector") - unname(fitted(object))
 }
+
+
+#' @describeIn mmrm_tmb_methods Simulates from an mmrm model object
+#' @importFrom stats simulate
+#' @exportS3Method
+#'
+#' @param object an object of class 'mmrm'
+#' @param nsim number of response vectors to simulate. Defaults to 1.
+#' @param seed an object specifying if and how the random number generator should be initialized (‘seeded’).
+#' @param newdata data frame for which to evaluate predictions.
+#' @param ...
+#'
+#' @return A \code{data.frame} of dimension [n, m] where n is the number of observations
+#' fitted in the model, and m is the number \code{nsim} of simulated responses.
+simulate.mmrm_tmb <- function(object, nsim = 1,
+                          seed = NULL, newdata = object$data,
+                          ...){
+  if(!is.null(seed)) set.seed(seed)
+
+  # build data.frame of prediction vectors
+  ret <- predict(object, newdata = newdata, se.fit = FALSE, interval = "none")
+  ret <- data.frame(replicate(nsim, ret))
+  covmat <- VarCorr(object)
+  grouped <- if(object$tmb_data$n_groups > 1) TRUE else FALSE
+  subjects <- unique(newdata$subject) # TODO
+  n_subjects <- length(subjects) # TODO
+  subject_n_visits <- apply(1:n_sujbects, 1, function(x) sum(subjects[subjects == x])) # TODO
+  subject_zero_inds <- apply(1:n_subjects, 1, function(x) first(subjects[subjects == x])) # TODO get zero inds
+  visit_zero_inds <- # TODO get visits
+
+  for(i in 1:n_subjects){
+
+    # Obtain indices of data.frame belonging to subject i
+    num_visits <- subject_n_visits[i]
+    inds <- (subject_zero_inds[i] + 1) : (subject_zero_inds[i] + num_visits)
```

Comment:
so the same issue here, if subject is not ordered, with (subject_zero_inds[i] + 1) : (subject_zero_inds[i] + num_visits) you will not get the records for that subject